# **Tarea Nº 3**
Scope: If, Loops, Functions

Curso:          Fundamentos de Programación en Python para Macroeconomía y Finanzas     
Profesor:       Victor Fuentes Campos     

- Fecha y hora de entrega: Miércoles 10 de enero, 11:59am
- Agregue los nombres de las personas con las que discutió esta tarea: ____ Luis Cahuana ______
- Envíe su tarea haciendo el `push` de su código a su repo en GitHub Classroom: 

As of 2022 Q2, there are almost 3.5 million Andriod apps on Google Play and 1.6 million iOS apps on the App Store, and counting. Giving the digital revolution, there is an increase interest in investment opportunities in this sector

<img src="popular_app_stores.png" max-width="300">

Under this context, you will assess a financial committee interested in investing in software development. The goal is to determine the kinds of apps that are likely to attract more users because the number of people using our apps affect our revenue. For this purpose, you will access to data from Google Play and App Store.

- Google Play: you can dowload the dataset containing approximately 11k Android apps from [this link](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv). It was collected in August 2018
- App Store: you can dowload the dataset containing roughly 7k iOS apps from [this link](https://dq-content.s3.amazonaws.com/350/AppleStore.csv). It was collected in July 2017

## Exploring data

1. Open the two datasets we mentioned above, and save both as lists of lists

In [3]:
# Solution
import pandas as pd

# Abrimos el primer conjunto de datos y lo guarda como una lista de listas
Google_Play = pd.read_csv(r"googleplaystore.csv") #(Colocar dirección del archivo csv)
dataset1 = Google_Play.values.tolist()

# Abrimos el segundo conjunto de datos y lo guarda como una lista de listas
Apple_Store = pd.read_csv(r"AppleStore.csv") #(Colocar dirección del archivo csv)
dataset2 = Apple_Store.values.tolist()

2. Print the first few rows of each dataset and find the number of rows and columns of each dataset

In [5]:
# Solution
Google_Play.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [8]:
print(f'Datos del primer conjunto {Google_Play.shape[0]} filas y {Google_Play.shape[1]} columnas.')

Datos del primer conjunto 10841 filas y 13 columnas.


For more info about the columns in the App Store dataset, visit [this link](https://www.kaggle.com/datasets/ramamet4/app-store-apple-data-set-10k-apps?select=AppleStore.csv)

## Preparing data

3. As noted before in the [discussion forum](https://www.kaggle.com/datasets/lava18/google-play-store-apps/discussion/66015), there is an error in the Google Play dataset. Please, remove the entry 10472 and explain shortly why

In [11]:
# Solution
print(Google_Play.iloc[10472])
Google_Play_sin_10472=Google_Play.drop(10472)
Google_Play_sin_10472.iloc[10472]

App               Life Made WI-Fi Touchscreen Photo Frame
Category                                              1.9
Rating                                               19.0
Reviews                                              3.0M
Size                                               1,000+
Installs                                             Free
Type                                                    0
Price                                            Everyone
Content Rating                                        NaN
Genres                                  February 11, 2018
Last Updated                                       1.0.19
Current Ver                                    4.0 and up
Android Ver                                           NaN
Name: 10472, dtype: object


App               osmino Wi-Fi: free WiFi
Category                            TOOLS
Rating                                4.2
Reviews                            134203
Size                                 4.1M
Installs                      10,000,000+
Type                                 Free
Price                                   0
Content Rating                   Everyone
Genres                              Tools
Last Updated               August 7, 2018
Current Ver                       6.06.14
Android Ver                    4.4 and up
Name: 10473, dtype: object

4. Count the number of duplicates in the Google Play dataset. You must build a function to loop through the Google Play dataset

In [17]:
# Solution
def contar_duplicados(Google_Play_sin_10472):
    total_filas=len(Google_Play_sin_10472)
    filas_únicas=len(Google_Play_sin_10472.drop_duplicates())
    filas_duplicadas = total_filas - filas_únicas
    return filas_duplicadas 

duplicados = contar_duplicados(Google_Play_sin_10472)
print( f'{duplicados} valores duplicados' )

483 valores duplicados


5. Remove duplicates ...

In [25]:
def eliminar_duplicados(Google_Play_sin_10472):
    
    Diccionario_sin_duplicados = Google_Play_sin_10472.drop_duplicates()
    return Diccionario_sin_duplicados


Diccionario_sin_duplicados=eliminar_duplicados(Google_Play_sin_10472)

5. a. Create a dictionary, where each dictionary key is a unique app name and the corresponding dictionary value is the highest number of reviews of that app


In [26]:
# Solution
def create_app_reviews_dictionary(dataframe):
    
    app_reviews_dict = {}

    for index, row in dataframe.iterrows():
        app_name = row['App']
        app_reviews = row['Reviews']

        if app_name not in app_reviews_dict or app_reviews > app_reviews_dict[app_name]:
            app_reviews_dict[app_name] = app_reviews

    return app_reviews_dict
result_dict = create_app_reviews_dictionary(Diccionario_sin_duplicados)


5. b. Use the information stored in the dictionary and create a new dataset, which will have only one entry per app (and for each app, we'll only select the entry with the highest number of reviews).


In [27]:
# Solution
Google_Play_filtered_dataframe = (Diccionario_sin_duplicados.sort_values(by='Reviews', ascending=False)
  .drop_duplicates('App')
 .reset_index(drop=True))

5. c. Explore the cleaned dataset, it should have 9,659 rows

In [28]:
# Solution
print(Google_Play_filtered_dataframe)

                                                    App          Category  \
0                                GollerCepte Live Score            SPORTS   
1                          Ad Block REMOVER - NEED ROOT             TOOLS   
2                                   SnipSnap Coupon App          SHOPPING   
3                     US Open Tennis Championships 2018            SPORTS   
4                                            DreamTrips  TRAVEL_AND_LOCAL   
...                                                 ...               ...   
9654  Curso Básico y Avanzado la Fe de Jesús Audio-T...            FAMILY   
9655                           CT Pulmonary Angiography           MEDICAL   
9656                           Theme For Galaxy Note FE   PERSONALIZATION   
9657                                     Toronto Dating            DATING   
9658                               PlayMotiv dn edition          BUSINESS   

      Rating Reviews  Size    Installs  Type Price Content Rating  \
0     

6. Remove non-english apps from both Android and Apple datasets. You should write a function that takes a string, if there's more than three characters that fall outside the ASCII range (0 - 127), then the function should return False (identify the string as non-English), otherwise it should return True. Hint: use the [```ord()```  function](https://www.w3schools.com/python/ref_func_ord.asp)

In [13]:
# For instance
correct_english_name = "Instachat 😜"  # 1 character outside the ASCII range (😜)
incorrect_english_name = "爱奇艺PPS -《欢乐颂2》电视剧热播" # 13 character outside the ASCII range (ie 播)

In [14]:
print(ord("I"))
print(ord("t"))
print(ord("-"))
print(ord("😜"))
print(ord("爱"))

73
116
45
128540
29233


In [ ]:
#Caso Google play

In [32]:
# Solution
Google_Play_filtered_dataframe.head(0)

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

In [36]:
def identificar_apps_en_ingles(Google_Play_filtered_dataframe, App):
    
    def en_ingles(App):
        caracteres_no_ascii = 0
        for caracter in App:
            if ord(caracter) > 127:
                caracteres_no_ascii += 1
        if caracteres_no_ascii > 3:
            return False
        else:
            return True

    Google_Play_filtered_dataframe.loc[:, 'En_ingles'] = Google_Play_filtered_dataframe[App].apply(en_ingles)
    return Google_Play_filtered_dataframe

nombres_de_aplicaciones_columna = 'App' 
Google_Play_apps = identificar_apps_en_ingles(Google_Play_filtered_dataframe, nombres_de_aplicaciones_columna)


Google_Play_apps

App          Category  \
0                                GollerCepte Live Score            SPORTS   
1                          Ad Block REMOVER - NEED ROOT             TOOLS   
2                                   SnipSnap Coupon App          SHOPPING   
3                     US Open Tennis Championships 2018            SPORTS   
4                                            DreamTrips  TRAVEL_AND_LOCAL   
...                                                 ...               ...   
9654  Curso Básico y Avanzado la Fe de Jesús Audio-T...            FAMILY   
9655                           CT Pulmonary Angiography           MEDICAL   
9656                           Theme For Galaxy Note FE   PERSONALIZATION   
9657                                     Toronto Dating            DATING   
9658                               PlayMotiv dn edition          BUSINESS   

      Rating Reviews  Size    Installs  Type Price Content Rating  \
0        4.2    9992   31M  1,000,000+  Free     0       Everyone   
1        3.3     999   91k    100,000+  Free     0       Everyone   
2        4.2    9975   18M  1,000,000+  Free     0       Everyone   
3        4.0    9971   33M  1,000,000+  Free     0       Everyone   
4        4.7    9971   22M    500,000+  Free     0           Teen   
...      ...     ...   ...         ...   ...   ...            ...   
9654     NaN       0   11M         10+  Free     0   Everyone 10+   
9655     NaN       0  4.9M        100+  Free     0       Everyone   
9656     NaN       0  7.5M        100+  Free     0       Everyone   
9657     NaN       0   14M        100+  Free     0     Mature 17+   
9658     NaN       0   28M        500+  Free     0       Everyone   

               Genres       Last Updated Current Ver Android Ver  En_ingles  
0              Sports       May 23, 2018         6.5  4.1 and up       True  
1               Tools  December 17, 2013         3.2  2.2 and up       True  
2            Shopping   January 22, 2018         1.4  4.3 and up       True  
3              Sports       June 5, 2018         7.1  5.0 and up       True  
4      Travel & Local     August 6, 2018      1.28.1  5.0 and up       True  
...               ...                ...         ...         ...        ...  
9654        Education      July 12, 2018         1.0  4.0 and up       True  
9655          Medical     March 24, 2018         1.0  4.0 and up       True  
9656  Personalization     April 19, 2018         1.1  4.1 and up       True  
9657           Dating      July 12, 2018         1.0  4.1 and up       True  
9658         Business      April 8, 2018       1.0.2  4.1 and up       True  

[9659 rows x 14 columns]

In [35]:
#Base de datos solo en ingles
Google_Play_ingles = Google_Play_apps[Google_Play_apps['En_ingles'] == True]
Google_Play_ingles

App          Category  \
0                                GollerCepte Live Score            SPORTS   
1                          Ad Block REMOVER - NEED ROOT             TOOLS   
2                                   SnipSnap Coupon App          SHOPPING   
3                     US Open Tennis Championships 2018            SPORTS   
4                                            DreamTrips  TRAVEL_AND_LOCAL   
...                                                 ...               ...   
9654  Curso Básico y Avanzado la Fe de Jesús Audio-T...            FAMILY   
9655                           CT Pulmonary Angiography           MEDICAL   
9656                           Theme For Galaxy Note FE   PERSONALIZATION   
9657                                     Toronto Dating            DATING   
9658                               PlayMotiv dn edition          BUSINESS   

      Rating Reviews  Size    Installs  Type Price Content Rating  \
0        4.2    9992   31M  1,000,000+  Free     0       Everyone   
1        3.3     999   91k    100,000+  Free     0       Everyone   
2        4.2    9975   18M  1,000,000+  Free     0       Everyone   
3        4.0    9971   33M  1,000,000+  Free     0       Everyone   
4        4.7    9971   22M    500,000+  Free     0           Teen   
...      ...     ...   ...         ...   ...   ...            ...   
9654     NaN       0   11M         10+  Free     0   Everyone 10+   
9655     NaN       0  4.9M        100+  Free     0       Everyone   
9656     NaN       0  7.5M        100+  Free     0       Everyone   
9657     NaN       0   14M        100+  Free     0     Mature 17+   
9658     NaN       0   28M        500+  Free     0       Everyone   

               Genres       Last Updated Current Ver Android Ver  En_ingles  
0              Sports       May 23, 2018         6.5  4.1 and up       True  
1               Tools  December 17, 2013         3.2  2.2 and up       True  
2            Shopping   January 22, 2018         1.4  4.3 and up       True  
3              Sports       June 5, 2018         7.1  5.0 and up       True  
4      Travel & Local     August 6, 2018      1.28.1  5.0 and up       True  
...               ...                ...         ...         ...        ...  
9654        Education      July 12, 2018         1.0  4.0 and up       True  
9655          Medical     March 24, 2018         1.0  4.0 and up       True  
9656  Personalization     April 19, 2018         1.1  4.1 and up       True  
9657           Dating      July 12, 2018         1.0  4.1 and up       True  
9658         Business      April 8, 2018       1.0.2  4.1 and up       True  

[9614 rows x 14 columns]

Caso Apple Store

In [31]:
Apple_Store.head(0)

Empty DataFrame
Columns: [id, track_name, size_bytes, currency, price, rating_count_tot, rating_count_ver, user_rating, user_rating_ver, ver, cont_rating, prime_genre, sup_devices.num, ipadSc_urls.num, lang.num, vpp_lic]
Index: []

In [37]:
def identificar_apps_en_ingles(Apple_Store, track_name):
    
    def en_ingles(track_name):
        caracteres_no_ascii = 0
        for caracter in track_name:
            if ord(caracter) > 127:
                caracteres_no_ascii += 1
        if caracteres_no_ascii > 3:
            return False
        else:
            return True

    Apple_Store.loc[:, 'En_ingles'] = Apple_Store[track_name].apply(en_ingles)
    return Apple_Store

nombres_de_aplicaciones_columna = 'track_name' 
Apple_Store_apps = identificar_apps_en_ingles(Apple_Store, nombres_de_aplicaciones_columna)

Apple_Store_apps

id                                    track_name  size_bytes  \
0      284882215                                      Facebook   389879808   
1      389801252                                     Instagram   113954816   
2      529479190                                Clash of Clans   116476928   
3      420009108                                    Temple Run    65921024   
4      284035177                       Pandora - Music & Radio   130242560   
...          ...                                           ...         ...   
7192  1170406182  Shark Boom - Challenge Friends with your Pet   245415936   
7193  1069830936                           【謎解き】ヤミすぎ彼女からのメッセージ    16808960   
7194  1070052833                                    Go!Go!Cat!    91468800   
7195  1081295232   Suppin Detective: Expose their true visage!    83026944   
7196   977965019           みんなのお弁当 by クックパッド ~お弁当をレシピ付きで記録・共有~    51174400   

     currency  price  rating_count_tot  rating_count_ver  user_rating  \
0         USD    0.0           2974676               212          3.5   
1         USD    0.0           2161558              1289          4.5   
2         USD    0.0           2130805               579          4.5   
3         USD    0.0           1724546              3842          4.5   
4         USD    0.0           1126879              3594          4.0   
...       ...    ...               ...               ...          ...   
7192      USD    0.0                 0                 0          0.0   
7193      USD    0.0                 0                 0          0.0   
7194      USD    0.0                 0                 0          0.0   
7195      USD    0.0                 0                 0          0.0   
7196      USD    0.0                 0                 0          0.0   

      user_rating_ver      ver cont_rating        prime_genre  \
0                 3.5     95.0          4+  Social Networking   
1                 4.0    10.23         12+      Photo & Video   
2                 4.5  9.24.12          9+              Games   
3                 4.0    1.6.2          9+              Games   
4                 4.5    8.4.1         12+              Music   
...               ...      ...         ...                ...   
7192              0.0    1.0.9          4+              Games   
7193              0.0      1.2          9+               Book   
7194              0.0    1.1.2         12+              Games   
7195              0.0    1.0.3         12+      Entertainment   
7196              0.0    1.4.0          4+       Food & Drink   

      sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  En_ingles  
0                  37                1        29        1       True  
1                  37                0        29        1       True  
2                  38                5        18        1       True  
3                  40                5         1        1       True  
4                  37                4         1        1       True  
...               ...              ...       ...      ...        ...  
7192               38                5         1        1       True  
7193               38                0         1        1      False  
7194               37                2         2        1       True  
7195               40                0         1        1       True  
7196               37                0         1        1      False  

[7197 rows x 17 columns]

In [39]:
Apple_Store_ingles = Apple_Store_apps[Apple_Store_apps['En_ingles'] == True]
Apple_Store_ingles

id                                    track_name  size_bytes  \
0      284882215                                      Facebook   389879808   
1      389801252                                     Instagram   113954816   
2      529479190                                Clash of Clans   116476928   
3      420009108                                    Temple Run    65921024   
4      284035177                       Pandora - Music & Radio   130242560   
...          ...                                           ...         ...   
7182  1070854722                                  Be-be-bears!   480781312   
7186  1169971902                   Hey Duggee: We Love Animals   136347648   
7192  1170406182  Shark Boom - Challenge Friends with your Pet   245415936   
7194  1070052833                                    Go!Go!Cat!    91468800   
7195  1081295232   Suppin Detective: Expose their true visage!    83026944   

     currency  price  rating_count_tot  rating_count_ver  user_rating  \
0         USD   0.00           2974676               212          3.5   
1         USD   0.00           2161558              1289          4.5   
2         USD   0.00           2130805               579          4.5   
3         USD   0.00           1724546              3842          4.5   
4         USD   0.00           1126879              3594          4.0   
...       ...    ...               ...               ...          ...   
7182      USD   2.99                 0                 0          0.0   
7186      USD   2.99                 0                 0          0.0   
7192      USD   0.00                 0                 0          0.0   
7194      USD   0.00                 0                 0          0.0   
7195      USD   0.00                 0                 0          0.0   

      user_rating_ver      ver cont_rating        prime_genre  \
0                 3.5     95.0          4+  Social Networking   
1                 4.0    10.23         12+      Photo & Video   
2                 4.5  9.24.12          9+              Games   
3                 4.0    1.6.2          9+              Games   
4                 4.5    8.4.1         12+              Music   
...               ...      ...         ...                ...   
7182              0.0  1.0.2.5          4+              Games   
7186              0.0      1.2          4+              Games   
7192              0.0    1.0.9          4+              Games   
7194              0.0    1.1.2         12+              Games   
7195              0.0    1.0.3         12+      Entertainment   

      sup_devices.num  ipadSc_urls.num  lang.num  vpp_lic  En_ingles  
0                  37                1        29        1       True  
1                  37                0        29        1       True  
2                  38                5        18        1       True  
3                  40                5         1        1       True  
4                  37                4         1        1       True  
...               ...              ...       ...      ...        ...  
7182               35                5        13        1       True  
7186               40                5         1        1       True  
7192               38                5         1        1       True  
7194               37                2         2        1       True  
7195               40                0         1        1       True  

[6183 rows x 17 columns]

7. Loop through each dataset to isolate the free apps in separate lists. Make sure you identify the columns describing the app price correctly. Be careful, prices come up as strings ('0', $0.99, $2.99, etc.), so make sure you're not checking an integer or a float in your conditional statements. How many apps are free?


In [43]:
# Solution
import pandas as pd
Google_Play_ingles.head(0)

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver, En_ingles]
Index: []

In [51]:
apps_gratuitas = Google_Play_ingles.loc[Google_Play_ingles['Price'] == '0']

lista_apps_gratuitas = apps_gratuitas['App'].tolist()

print(f'En Google Play hay {len(lista_apps_gratuitas)} aplicaciones gratuitas.')

En Google Play hay 8862 aplicaciones gratuitas.


In [52]:
Apple_Store_ingles.head(0)

Empty DataFrame
Columns: [id, track_name, size_bytes, currency, price, rating_count_tot, rating_count_ver, user_rating, user_rating_ver, ver, cont_rating, prime_genre, sup_devices.num, ipadSc_urls.num, lang.num, vpp_lic, En_ingles]
Index: []

In [55]:
apps_gratuitas2 = Apple_Store_ingles.loc[Apple_Store_ingles['price'] == 0]

lista_apps_gratuitas2=apps_gratuitas2['track_name'].tolist()

print(f"En el caso de Apple Store hay {len(lista_apps_gratuitas2)} aplicaciones gratuitas.")

En el caso de Apple Store hay 3222 aplicaciones gratuitas.


## Exploring data

 8. Build a frequency table for the ```prime_genre``` column of the App Store dataset, and the ```Genres``` and ```Category``` columns of the Google Play data set. What are the most common genres for each market? You must develop and deploy your function

In [57]:
# Solution

def frequency_table(dataframe, column_name):
    return dataframe[column_name].value_counts().reset_index().rename(columns={'index': column_name, column_name: 'Frequency'})

# Frecuencias de 'prime_genre'
app_store_frequency = frequency_table(apps_gratuitas2, 'prime_genre')

# Frecuencias de 'Genres'
google_play_genres_frequency = frequency_table(apps_gratuitas, 'Genres')

# Frecuencias de 'Category'
google_play_category_frequency = frequency_table(apps_gratuitas, 'Category')
print(app_store_frequency.head())
print(google_play_genres_frequency.head())
print(google_play_category_frequency.head())

         prime_genre  Frequency
0              Games       1874
1      Entertainment        254
2      Photo & Video        160
3          Education        118
4  Social Networking        106
          Genres  Frequency
0          Tools        748
1  Entertainment        538
2      Education        474
3       Business        407
4      Lifestyle        345
    Category  Frequency
0     FAMILY       1680
1       GAME        859
2      TOOLS        749
3   BUSINESS        407
4  LIFESTYLE        346


 9. One way to find out what genres are the most popular (have the most users) is to calculate the average number of installs for each app genre. Based on that exercise, give one app profile recommendation for Google Play.
 
    For the Google Play data set, we can find the required information in the ```Installs``` column. Then, compute the average number of installs by genre. Hint 1: consider only unique values of the ```Category``` column. Hint 2: If an app has, for instance, ```1,000,000+``` assume it as ```1,000,000``` installs. Hint 3: Be careful with a skewed distribution

In [65]:
# Solution
instalaciones_promedio_por_genero = apps_gratuitas.groupby('Category')('Installs').mean().reset_index()
instalaciones_promedio_por_genero = instalaciones_promedio_por_genero.sort_values(by='Installs', ascending=False)

print(instalaciones_promedio_por_genero.head(5))  

TypeError: 'DataFrameGroupBy' object is not callable

 10. The App Store dataset doesn´t have the ```installs``` information. Then, you will construct the average rating from the ```rating_count_tot``` column per genre. Based on that exercise, give one app profile recommendation for the App Store

In [ ]:
# Solution